[Hrishiraj Mandal] ip

Virusrwj223 opened the pull request:
# Mendel

> "There is no place like localhost" - [Mandal](https://www.linkedin.com/in/hrishiraj-mandal)

Mendel is straight forward and to the point chat-based task reminder utility app. It's
- text-based
- easy to learn
- *Blazing* fast usage :fire: 

All you need to do is,
1. Download it from [here](https://github.com/Virusrwj223/ip/releases/download/A-Jar/mendel-1-a.jar).
2. Copy the jar file into an empty folder.
3. Open a command window in that folder.
4. run the command `java -jar "mendel-1-a.jar"` in the same folder as the jar file.
5. Add your tasks.
6. Let it manage your tasks.

Features:
- [x] add Todo, Deadline, Event
- [x] find tasks based on matching description
- [x] delete tasks
- [x] mark and unmark tasks
- [x] list tasks
- [x] auto save persistently 

For developers:
The `main` method is as such:
```
public class Mendel {
    private final Parser taskManager;
    private final UI ui;

    public Mendel() {
        this.taskManager = new Parser();
        this.ui = new UI();
    }

    public void run() {
        Scanner sc = new Scanner(System.in);
        String currAction = "hello";

        ui.showWelcome();
        while (!currAction.equals("bye")) {
            currAction = sc.nextLine().trim();
            try {
                ui.preetyPrint(taskManager.manage(currAction));
            } catch (MendelException e) {
                ui.preetyPrint(e.toString());
            } catch (ServerError e) {
                ui.preetyPrint(e.toString());
                taskManager.manage("bye");
                currAction = "bye";
            }
        }
    }

    public static void main(String[] args) {
        new Mendel().run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Virusrwj223 committed:
Rename Duke to Mendel

Virusrwj223 committed:
Add Mendel file

Virusrwj223 committed:
Rename Duke placeholders

Virusrwj223 committed:
Add Welcome feature

Virusrwj223 committed:
Add Leave feature

Virusrwj223 committed:
Add indent format

Virusrwj223 committed:
Add echo task feature

Virusrwj223 committed:
enable hello echo

Virusrwj223 committed:
add task storage feature

Virusrwj223 committed:
Refactor add to Task class

Virusrwj223 committed:
Add mark feature

Virusrwj223 committed:
Add specific adding features

Virusrwj223 committed:
Refactor command response

Virusrwj223 committed:
Setup testing

Virusrwj223 committed:
Catch unknownCmdException

Virusrwj223 committed:
Catch input exceptions

Virusrwj223 committed:
Handle marking errors

Virusrwj223 committed:
Add delete feature

Virusrwj223 committed:
Refactor text format to package

Virusrwj223 committed:
Refactor task and subclasses

Virusrwj223 committed:
Refactor MendelException

Virusrwj223 committed:
Refactor metacognitive abilities

Virusrwj223 committed:
Refactor add taskStorage method

Virusrwj223 committed:
Wrap exception handling

Virusrwj223 committed:
Task: Make abstract

Virusrwj223 committed:
Refactor task

Virusrwj223 committed:
DB: read access

Virusrwj223 committed:
Refactor tasks for db read

Virusrwj223 committed:
DB: write access

Virusrwj223 committed:
Handle Internal Server Error

Virusrwj223 committed:
Merge branch 'branch-Level-7'

Virusrwj223 committed:
Merge pull request #1 from Virusrwj223/branch-Level-7

Branch level 7

Virusrwj223 committed:
Update Mendel

Virusrwj223 committed:
Merge branch 'master' of https://github.com/Virusrwj223/ip

Virusrwj223 committed:
Identify dateObject in raw String

Virusrwj223 committed:
Feature: findEventOn

Virusrwj223 committed:
Develop dateTime feature

Virusrwj223 committed:
Refactor UI welcome

Virusrwj223 committed:
Refactor preetyPrint

Virusrwj223 committed:
Refactor Storage

Virusrwj223 committed:
Refactor Parser

Virusrwj223 committed:
Refactor TaskList

Virusrwj223 committed:
BugFix: same error msg

Virusrwj223 committed:
Refactor old class name to assignment class name

Virusrwj223 committed:
Refactor frontend package

Virusrwj223 committed:
Refactor package

Virusrwj223 committed:
Merge branch 'add-gradle-support'

Virusrwj223 committed:
Test Date string

Virusrwj223 committed:
Test datetime class

Virusrwj223 committed:
Test deadline

Virusrwj223 committed:
Merge branch 'test'

Virusrwj223 committed:
Update gradle build config

Virusrwj223 committed:
Merge branch 'production'

Virusrwj223 committed:
Document DateTimeManager

Virusrwj223 committed:
Document Storage

Virusrwj223 committed:
Document Deadline

Virusrwj223 committed:
Document Event

Virusrwj223 committed:
Document Todo

Virusrwj223 committed:
Document Task

Virusrwj223 committed:
Document Parser

Virusrwj223 committed:
Document Parser

Virusrwj223 committed:
Document Conditional Exception Handler

Virusrwj223 committed:
Document TaskList

Virusrwj223 committed:
Merge branch 'branch-A-JavaDoc'

Virusrwj223 committed:
Remove dead code

Virusrwj223 committed:
Merge branch 'branch-A-CodingStandard'

Virusrwj223 committed:
Feature find

Virusrwj223 committed:
Merge branch 'branch-Level-9'

Virusrwj223 committed:
Move main to main package

Virusrwj223 committed:
Refactor main

Virusrwj223 committed:
Update gradle main location

Virusrwj223 committed:
Merge branch 'refactor'

hooyushing commented on src/main/java/mendel/datetime/DateTimeManager.java:
```diff
@@ -0,0 +1,98 @@
+            this.formattedDate = date.format(DateTimeFormatter.ofPattern("MMM dd yyyy"));
+        } catch (DateTimeParseException e) {
+            boolean hasValidFormat = false;
+            String[] possibleFormatsTimed = new String[] {"dd-MMM-yyyy HH:mm", "dd/MMM/yyyy HH:mm", "dd MMM yyyy HH:mm",
```
Is it possible to indent a new line so the code follows the coding standard?

hooyushing commented on src/main/java/mendel/datetime/DateTimeManager.java:
```diff
@@ -0,0 +1,98 @@
+                    "d-M-yyyy HH mm", "d-M-yyyy HHmm", "d-MMM-yyyy HH mm", "d-MMM-yyyy HHmm", "d MMM yyyy HHmm"};
+            for (int i = 0; i < possibleFormatsTimed.length; i++) {
+                if (this.isValidFormat(rawDate, DateTimeFormatter.ofPattern(possibleFormatsTimed[i]))) {
+                    LocalDateTime date = LocalDateTime.parse(rawDate, DateTimeFormatter.ofPattern(possibleFormatsTimed[i]));
```
Is it possible to indent new line so the code follows the coding standard and each line is less than 120 characters?

hooyushing commented on src/main/java/mendel/dbmanager/Storage.java:
```diff
@@ -0,0 +1,124 @@
+     * @param taskStorage the TaskList object to load tasks into.
+     * @throws ServerError if the file is not found or if the task type in the file is unidentifiable.
+     */
+    public void loadInto(TaskList taskStorage) throws ServerError {
```
Is it possible to change the name of the error so it follows the coding standard? i.e. loadError

hooyushing commented on src/main/java/mendel/dbmanager/Storage.java:
```diff
@@ -0,0 +1,124 @@
+     * @param isNew a boolean indicating whether the database is new created to avoid formatting error.
+     * @throws ServerError if the file cannot be found or an IO error occurs during writing.
+     */
+    public void create(Task task, boolean isNew) throws ServerError {
```
Is it possible to change the name of the error so it follows the coding standard? i.e. createError

hooyushing commented on src/main/java/mendel/discretetask/Deadline.java:
```diff
@@ -0,0 +1,141 @@
+                        "OOPS! I am unsure of due.\nPlease specify only one due.");
+        String endMsg = slashSegments[1];
+        ConditionalExceptionHandler.of()
+                .conditionTriggerException(endMsg.isEmpty(), "OOPS! I am unsure of due.\nPlease specify a due.");
```
Is it possible to break the line up so it follows the coding standard of less than 120 characters per line?

hooyushing commented on src/main/java/mendel/discretetask/Deadline.java:
```diff
@@ -0,0 +1,141 @@
+     * @param rawDescription The raw description string containing the task description and due date.
+     * @return A new Deadline object with the parsed description and due date.
+     */
+    public static Deadline of(String rawDescription) {
```
I like how you created a class factory method!

hooyushing commented on src/main/java/mendel/discretetask/Todo.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    /**
+     * Checks if the given date matches the due date for this Todo task.
```
I like how you add JavaDoc comments to majority of the methods so it is easier to understand!

hooyushing commented on src/main/java/mendel/frontend/Parser.java:
```diff
@@ -0,0 +1,86 @@
+     * @return The response message after executing the command.
+     * @throws MendelException If the command is invalid or if there is an error processing it.
+     */
+    public String manage(String currAction) throws MendelException {
```
Is it possible you could create a manageException to follow the coding standard? :D

hooyushing commented on src/main/java/mendel/dbmanager/Storage.java:
```diff
@@ -0,0 +1,124 @@
+                String line = s.nextLine();
+                String[] lineSegments = line.split(" \\| ");
+                if (lineSegments[0].equals("T")) {
+                    taskStorage.silencedAdd(Todo.loadOf(lineSegments[1].equals("1"), lineSegments[2]));
```
Is it possible to place breaks to improve readability?

hooyushing commented on src/main/java/mendel/datetime/DateTimeManager.java:
```diff
@@ -0,0 +1,98 @@
+                    "M d yyyy", "d M yyyy", "d/M/yyyy", "M/d/yyyy", "yyyy d M", "yyyy, d M", "d-M-yyyy"};
+            for (int i = 0; i < possibleFormatsUnTimed.length; i++) {
+                if (this.isValidFormat(rawDate, DateTimeFormatter.ofPattern(possibleFormatsUnTimed[i]))) {
+                    LocalDate date = LocalDate.parse(rawDate, DateTimeFormatter.ofPattern(possibleFormatsUnTimed[i]));
```
Is it possible to indent a new line so the code follows the coding standard?

hooyushing commented on src/main/java/mendel/datetime/DateTimeManager.java:
```diff
@@ -0,0 +1,98 @@
+     */
+    public String removeTimeStamp() {
+        try {
+            LocalDate date = LocalDate.parse(this.formattedDate, DateTimeFormatter.ofPattern("MMM dd yyyy, HH:mm"));
```
Is it possible to indent a new line so the code follows the coding standard?

hooyushing commented on src/main/java/mendel/dbmanager/Storage.java:
```diff
@@ -0,0 +1,124 @@
+                } else if (lineSegments[0].equals("D")) {
+                    taskStorage.silencedAdd(Deadline.loadOf(lineSegments[1].equals("1"), lineSegments[2], lineSegments[3]));
+                } else if (lineSegments[0].equals("E")) {
+                    taskStorage.silencedAdd(Event.loadOf(lineSegments[1].equals("1"), lineSegments[2], lineSegments[3],
```
Is it possible to indent a new line so the code follows the coding standard?

hooyushing commented on src/main/java/mendel/dbmanager/Storage.java:
```diff
@@ -0,0 +1,124 @@
+                    taskStorage.silencedAdd(Event.loadOf(lineSegments[1].equals("1"), lineSegments[2], lineSegments[3],
+                            lineSegments[4]));
+                } else {
+                    throw new ServerError(String.format("Unidentifiable task type %s. Ensure correct task type", lineSegments[0]));
```
Is it possible to indent a new line so the code follows the coding standard?

hooyushing commented on src/main/java/mendel/dbmanager/Storage.java:
```diff
@@ -0,0 +1,124 @@
+package mendel.dbmanager;
```
I like how there are multiple packages to store different classes :D

Virusrwj223 committed:
Configure checkstyle

Virusrwj223 committed:
Fix style tests

Virusrwj223 committed:
Fix style error

Virusrwj223 committed:
Fix style error

Virusrwj223 committed:
Fix style error

Virusrwj223 committed:
Merge branch 'branch-A-CheckStyle'

Virusrwj223 committed:
Configure javafx

Virusrwj223 committed:
Style gui

Virusrwj223 committed:
Style gui chat format

Virusrwj223 committed:
Refactor fxml

Virusrwj223 committed:
Update gui cosmetic

Virusrwj223 committed:
Merge branch 'branch-Level-10'

Virusrwj223 committed:
Delete image

Virusrwj223 committed:
Merge branch 'branch-Level-10'

hooyushing commented on src/main/java/mendel/discretetask/Event.java:
```diff
@@ -0,0 +1,162 @@
+     * @throws MendelException if there are issues such as missing description,
+     *                         missing start or end date, or incorrect formatting.
+     */
+    private static void handleError(String rawDescription) throws MendelException {
```
Is it possible to change this exception to handleException to adhere to the coding standard?

hooyushing reviewed (commented):
Just some minor coding violations but overall good job!

Virusrwj223 committed:
Implement system exit

Virusrwj223 committed:
Merge branch 'branch-Level-10'

Nigeltzy commented on src/main/java/mendel/dbmanager/Storage.java:
```diff
@@ -0,0 +1,126 @@
+     * Copies data from the database file into the TaskList.
+     * Each line is parsed into the appropriate Task object (Todo, Deadline, or Event).
+     *
+     * @param taskStorage the TaskList object to load tasks into.
+     * @throws ServerError if the file is not found or if the task type in the file is unidentifiable.
```
Perhaps the first letter (eg. 'the', 'if') could be capitalized as per the coding standard. Maybe the "the" could be omitted to make comments shorter without affecting readability.

Nigeltzy commented on src/main/java/mendel/dbmanager/Storage.java:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * Appends a new entry for a task in the database file.
+     *
+     * @param task the Task object to be added to the file.
+     * @param isNew a boolean indicating whether the database is new created to avoid formatting error.
+     * @throws ServerError if the file cannot be found or an IO error occurs during writing.
```
Similar JavaDocs comment as raised previously

Nigeltzy commented on src/main/java/mendel/discretetask/Deadline.java:
```diff
@@ -0,0 +1,142 @@
+ * and includes additional attributes to handle the description and the due date.
+ */
+public class Deadline extends Task {
+    private final String description;
+    private final String by;
```
I like how relevant concepts of immutability was applied here

Nigeltzy commented on src/main/java/mendel/discretetask/Task.java:
```diff
@@ -0,0 +1,82 @@
+     */
+    public Task(String description) {
+        this.description = description;
+        this.isDone = false;
```
While not in the [basic-intermediate] section of the coding standards, perhaps you could consider applying the "use this keyword only when a field is shadowed by a method or constructor parameter." idea under the advanced rule section of the guidelines.

Nigeltzy commented on src/main/java/mendel/frontend/UI.java:
```diff
@@ -0,0 +1,42 @@
+     * @return The wrapped message with borders.
+     */
+    public String wrapLines(String message) {
+        return "____________________________________________________________\n" + message.indent(3)
+                + "____________________________________________________________";
```
I like how you followed the "break before operator" guideline

Nigeltzy commented on src/main/java/mendel/metacognition/TaskList.java:
```diff
@@ -0,0 +1,212 @@
+     * @return A message indicating that the task has been marked as done.
+     * @throws MendelException If the serial number is out of bounds.
+     */
+    public String marker(int serial) throws MendelException {
```
As this is a method, perhaps this could be a verb (eg. mark) instead?

Nigeltzy reviewed (commented):
LGTM. Only minor cosmetic changes proposed

Virusrwj223 committed:
Merge branch 'master' into refactor

Virusrwj223 committed:
Refactor Launcher to main package

Virusrwj223 committed:
Merge branch 'refactor'

Virusrwj223 committed:
Merge branch 'master' into production

Virusrwj223 committed:
Create latest jar

Virusrwj223 committed:
Merge branch 'production'

Virusrwj223 committed:
Assert storage

Virusrwj223 committed:
Assert discrete tasks descriptions

Virusrwj223 committed:
Check assumptions

Assertions have been utilised to check for assumptions made in the code. This is typically used in checking array length that is not facing the user. This is because there could be errors made through external corruption. Since the code assumes that non-user facing data is correct, this assumption may not be the case if there exists external corruption. Thus, assertions have been used to guard against this possibility.

Virusrwj223 committed:
Merge branch 'master' into branch-A-CodeQuality

Virusrwj223 committed:
Refactor DateTimeManager

Virusrwj223 committed:
Refactor Storage

Virusrwj223 committed:
Refactor task

Virusrwj223 committed:
Refactor Parser

Virusrwj223 committed:
Refactor TaskList

Virusrwj223 committed:
Refactor project

The project has been refactored to adhere to coding standards. It also utilises software engineering principles as guidelines to improve code structure and maintain future code readability

Virusrwj223 committed:
Add Continuous Integration

Virusrwj223 committed:
Merge branch 'master' into branch-A-CodeQuality

Virusrwj223 committed:
Refactor DateTimeManager

Virusrwj223 committed:
Refactor DateTimeManager indent

Virusrwj223 committed:
Refactor whitespace and lining

Virusrwj223 committed:
Refactor yml script

Virusrwj223 committed:
Refactor project

The project has been refactored to adhere to coding standards. It also utilises software engineering principles as guidelines to improve code structure and maintain future code readability. The CI has also been integrated to ensure consistency of coding standards.

Virusrwj223 committed:
Refactor DateTimeManager

Virusrwj223 committed:
Command Reminder Feature

Virusrwj223 committed:
Add auto-remind deadline on boot

Virusrwj223 committed:
Merge branch 'branch-B-Reminders'

Virusrwj223 committed:
Add conditional monad

Virusrwj223 committed:
Remove Conditional monad

Virusrwj223 committed:
Merge branch 'branch-A-CodeQuality'

Virusrwj223 committed:
Set UI min size

Virusrwj223 committed:
Show welcome message

Virusrwj223 committed:
Dark theme UI

Virusrwj223 committed:
Revert "Dark theme UI"

This reverts commit cad2fa3e2f9229beb100d59fcd2df10920b3d66c.

Virusrwj223 committed:
Revert "Set UI min size"

This reverts commit 6b59e48993b3420be865575bbdd55aebee5cda3b.

Virusrwj223 committed:
Spell check

Virusrwj223 committed:
Merge branch 'A-BetterGui'

Virusrwj223 committed:
Add dark dialog

Virusrwj223 committed:
Make dark theme

Virusrwj223 committed:
Test Todo

Virusrwj223 committed:
Test Event

Virusrwj223 committed:
Merge branch 'branch-A-MoreTesting'

Virusrwj223 committed:
Fix lint errors

Virusrwj223 committed:
Update ReadMe

Virusrwj223 committed:
Limit window

Virusrwj223 committed:
Demo UI

Virusrwj223 committed:
Remove screenshot

Virusrwj223 committed:
Screenshot UI

Virusrwj223 committed:
Name main readme title

Virusrwj223 committed:
Remove title bar image

Virusrwj223 committed:
Remove product screenshot

Virusrwj223 committed:
Update product screenshot

Virusrwj223 committed:
Update markdown

Virusrwj223 committed:
Merge branch 'branch-UserGuide'

Virusrwj223 committed:
Produce Production

Virusrwj223 committed:
Merge branch 'master' into production

Virusrwj223 committed:
Merge branch 'production'

Virusrwj223 committed:
Setup guide in ReadMe

Virusrwj223 committed:
Add linebbreak

Virusrwj223 committed:
Format table

Virusrwj223 committed:
Edit setup path

Virusrwj223 committed:
Edit table

Virusrwj223 committed:
Edit table

Virusrwj223 committed:
Blank space table

Virusrwj223 committed:
Update table

Virusrwj223 committed:
Persist initial table

Virusrwj223 committed:
Update readme

Virusrwj223 committed:
Update ReadMe

Virusrwj223 committed:
Clarify remind

Virusrwj223 committed:
Add footnote

Virusrwj223 committed:
Redraft footnote

Virusrwj223 committed:
Handle unlogical date period

Virusrwj223 committed:
Bug fix date time handling

Virusrwj223 committed:
Fix remind date logic error

Virusrwj223 committed:
Fix remind for event

Virusrwj223 committed:
Merge branch 'branch-A-MoreErrorHandling'

Virusrwj223 committed:
Merge branch 'master' into branch-A-MoreTesting

Virusrwj223 committed:
Add javadoc

Virusrwj223 committed:
Support more time format

Virusrwj223 committed:
Test deadline datetime

Virusrwj223 committed:
Test event datetime

Virusrwj223 committed:
Test event datetime

Virusrwj223 committed:
Test tasklist

Virusrwj223 committed:
Merge branch 'branch-A-MoreTesting'

Virusrwj223 committed:
Fix lint

Virusrwj223 committed:
Merge branch 'master' into branch-A-MoreErrorHandling

Virusrwj223 committed:
Fix whitespace issue

Virusrwj223 committed:
Refactor code

Virusrwj223 committed:
Fix space error

Virusrwj223 committed:
Merge branch 'branch-A-MoreErrorHandling'

Virusrwj223 committed:
Fix command case issue

Virusrwj223 committed:
Merge branch 'branch-A-MoreErrorHandling'

Virusrwj223 committed:
Solve todo casing

Virusrwj223 committed:
Add notes

Virusrwj223 committed:
Fix javadoc spell

Virusrwj223 committed:
Add line break to footnote

Virusrwj223 committed:
Command case

Virusrwj223 committed:
Merge branch 'branch-A-MoreErrorHandling'

Virusrwj223 committed:
Fix testcase

Virusrwj223 committed:
Update docs date format

Virusrwj223 committed:
Update credits

Virusrwj223 committed:
Fix list error

Virusrwj223 committed:
Remove picture

Virusrwj223 committed:
Update credit

Virusrwj223 committed:
Declare AI
